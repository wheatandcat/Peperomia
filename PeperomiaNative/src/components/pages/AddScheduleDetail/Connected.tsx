import { SQLite } from "expo-sqlite";
import React, { Component } from "react";
import { NavigationScreenProp, NavigationRoute } from "react-navigation";
import uuidv1 from "uuid/v1";
import { db } from "../../../lib/db";
import {
  ItemDetailParam,
  insert as insertItemDetail,
  ItemDetail
} from "../../../lib/db/itemDetail";
import getKind from "../../../lib/getKind";
import { SuggestItem } from "../../../lib/suggest";
import { Consumer as ItemsConsumer } from "../../../containers/Items";
import Page from "../../templates/CreateScheduleDetail/Page";

export interface State extends ItemDetailParam {
  iconSelected: boolean;
  suggestList: SuggestItem[];
}

interface Props extends ItemDetailParam {
  navigation: NavigationScreenProp<NavigationRoute>;
}

interface PlanProps extends Props {
  itemDetails: ItemDetail[];
  refreshData: () => void;
}

export default class extends Component<Props> {
  render() {
    return (
      <ItemsConsumer>
        {({ refreshData, itemDetails }: any) => (
          <Plan
            {...this.props}
            refreshData={refreshData}
            itemDetails={itemDetails}
          />
        )}
      </ItemsConsumer>
    );
  }
}

class Plan extends Component<PlanProps, State> {
  state = {
    title: this.props.title || "",
    place: this.props.place || "",
    url: this.props.url || "",
    memo: this.props.memo || "",
    moveMinutes: this.props.moveMinutes || 0,
    kind: this.props.kind,
    priority: this.props.priority,
    iconSelected: false,
    suggestList: []
  };

  componentDidMount() {
    const suggestList = this.props.itemDetails.map(itemDetail => ({
      title: itemDetail.title,
      kind: itemDetail.kind
    }));

    this.setState({
      suggestList
    });
  }

  componentDidUpdate() {
    const kind = this.props.navigation.getParam("kind", "");

    if (!kind) {
      return;
    }

    if (this.state.kind !== kind) {
      this.setState({ kind, iconSelected: true });
    }
  }

  onDismiss = () => {
    this.props.navigation.goBack();
  };

  onSave = (
    title: string,
    kind: string,
    place: string,
    url: string,
    memo: string,
    time: number
  ) => {
    const itemId = this.props.navigation.getParam("itemId", "1");
    const priority = this.props.navigation.getParam("priority", "1");

    db.transaction((tx: SQLite.Transaction) => {
      const itemDetail: ItemDetail = {
        itemId,
        title,
        place,
        url,
        memo,
        kind,
        moveMinutes: time,
        priority: Number(priority)
      };

      insertItemDetail(tx, itemDetail, this.save);
    });
  };

  save = () => {
    const itemId = this.props.navigation.getParam("itemId", "1");

    this.props.navigation.navigate("Schedule", {
      itemId,
      refresh: uuidv1()
    });

    this.props.refreshData();
  };

  onIcons = (title: string) => {
    this.props.navigation.navigate("Icons", {
      kind: getKind(title),
      onSelectIcon: (kind: string) => {
        this.props.navigation.navigate("AddScheduleDetail", {
          kind: kind
        });
      },
      onDismiss: () => {
        this.props.navigation.navigate("AddScheduleDetail");
      },
      photo: false
    });
  };

  render() {
    return (
      <Page
        title={this.state.title}
        kind={this.state.kind}
        place={this.state.place}
        url={this.state.url}
        memo={this.state.memo}
        time={this.state.moveMinutes}
        suggestList={this.state.suggestList}
        iconSelected={this.state.iconSelected}
        onDismiss={this.onDismiss}
        onSave={this.onSave}
        onIcons={this.onIcons}
      />
    );
  }
}
