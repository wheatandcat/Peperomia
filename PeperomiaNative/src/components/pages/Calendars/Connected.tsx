import React, { Component } from 'react';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import { createStackNavigator } from 'react-navigation-stack';
import {
  Consumer as ItemsConsumer,
  ContextProps,
} from '../../../containers/Items';
import theme from '../../../config/theme';
import Schedule from '../Schedule/Switch';
import EditPlan from '../EditPlan/Connected';
import Page from './Page';

type Props = {
  navigation: NavigationScreenProp<NavigationRoute>;
};

class Container extends Component<Props> {
  static navigationOptions = () => {
    return {
      header: null,
    };
  };

  onCreate = (date: string) => {
    this.props.navigation.navigate('CreatePlan', {
      date,
    });
  };

  onSchedule = (id: number, title: string) => {
    this.props.navigation.navigate('Schedule', { itemId: id, title });
  };

  render() {
    return (
      <ItemsConsumer>
        {({ calendars, itemsLoading }: ContextProps) => (
          <Page
            calendars={calendars || []}
            loading={itemsLoading || false}
            onCreate={this.onCreate}
            onSchedule={this.onSchedule}
          />
        )}
      </ItemsConsumer>
    );
  }
}

const MainCardNavigator = createStackNavigator(
  {
    Container: {
      screen: Container,
    },
    Schedule: {
      screen: Schedule,
    },
  },
  {
    defaultNavigationOptions: {
      headerStyle: {
        backgroundColor: theme().mode.header.backgroundColor,
      },
      headerTitleStyle: {
        color: theme().mode.header.text,
      },
      headerTintColor: theme().mode.header.text,
    },
  }
);

export default createStackNavigator(
  {
    MainCardNavigator: {
      screen: MainCardNavigator,
    },
    EditPlan: {
      screen: EditPlan,
    },
  },
  {
    initialRouteName: 'MainCardNavigator',
    headerMode: 'none',
  }
);
