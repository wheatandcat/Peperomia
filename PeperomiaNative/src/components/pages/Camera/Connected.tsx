import React, { Component } from 'react';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import EStyleSheet from 'react-native-extended-stylesheet';
import { TouchableOpacity, View } from 'react-native';
import { MaterialCommunityIcons } from '@expo/vector-icons';
import Page, { Props as PageProps } from './Page';

interface Props extends PageProps {
  navigation: NavigationScreenProp<NavigationRoute>;
}

export default class extends Component<Props> {
  static navigationOptions = ({
    navigation,
  }: {
    navigation: NavigationScreenProp<NavigationRoute>;
  }) => {
    const { params = {} } = navigation.state;

    return {
      title: 'アイコン写真を取る',
      headerLeft: (
        <View style={styles.headerLeft}>
          <TouchableOpacity onPress={params.onDismiss}>
            <MaterialCommunityIcons name="close" size={25} />
          </TouchableOpacity>
        </View>
      ),
    };
  };

  componentDidMount() {
    const onDismiss = this.props.navigation.getParam('onDismiss', () => {});

    this.props.navigation.setParams({
      onDismiss: onDismiss,
    });
  }

  onPicture = (image?: string) => {
    const onPicture = this.props.navigation.getParam('onPicture', () => {});

    onPicture(image);
  };

  render() {
    return <Page onPicture={this.onPicture} />;
  }
}

const styles = EStyleSheet.create({
  headerLeft: {
    left: 10,
  },
});
