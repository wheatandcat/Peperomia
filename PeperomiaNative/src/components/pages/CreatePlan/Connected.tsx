import * as ImageManipulator from "expo-image-manipulator";
import * as SQLite from "expo-sqlite";
import React, { Component } from "react";
import { Alert } from "react-native";
import { NavigationScreenProp, NavigationRoute } from "react-navigation";
import {
  Consumer as ItemsConsumer,
  ContextProps
} from "../../../containers/Items";
import { db, ResultError } from "../../../lib/db";
import { insert as insertItem, Item } from "../../../lib/db/item";
import { insert as insertCalendar, Calendar } from "../../../lib/db/calendar";
import { SuggestItem } from "../../../lib/suggest";
import getKind from "../../../lib/getKind";
import Page from "../../templates/CreatePlan/Page";

interface Props {
  navigation: NavigationScreenProp<NavigationRoute>;
}

interface State {
  input: {
    title: string;
    date: string;
  };
  image: string;
  kind: string;
  suggestList: SuggestItem[];
}

export default class extends Component<Props> {
  render() {
    return (
      <ItemsConsumer>
        {({ items, refreshData, calendars }: ContextProps) => (
          <Connect
            {...this.props}
            items={items}
            refreshData={refreshData}
            calendars={calendars}
          />
        )}
      </ItemsConsumer>
    );
  }
}

type ConnectProps = Props &
  Pick<ContextProps, "items" | "refreshData" | "calendars">;

class Connect extends Component<ConnectProps, State> {
  state = {
    input: { title: "", date: "" },
    image: "",
    kind: "",
    suggestList: []
  };

  componentDidMount() {
    const date = this.props.navigation.getParam("date", "");

    const suggestList = (this.props.items || []).map(item => ({
      title: item.title,
      kind: item.kind
    }));

    this.setState({
      suggestList,
      input: {
        ...this.state.input,
        date
      }
    });
  }

  async componentDidUpdate() {
    const image = this.props.navigation.getParam("image", "");
    if (image && image !== this.state.image) {
      this.setState({
        image
      });
    }

    const kind = this.props.navigation.getParam("kind", "");

    if (kind && kind !== this.state.kind) {
      this.setState({
        kind
      });
    }
  }

  onImage = (image: string) => {
    this.setState({
      image
    });
  };

  onInput = (name: string, value: any) => {
    this.setState({
      input: {
        ...this.state.input,
        [name]: value
      }
    });
  };

  onSave = async () => {
    if (this.state.input.date) {
      const check = (this.props.calendars || []).find(
        calendar => calendar.date === this.state.input.date
      );

      if (check) {
        Alert.alert("同じ日にスケジュールが既に登録されています");
        return;
      }
    }

    let image = "";
    if (this.state.image) {
      const manipResult = await ImageManipulator.manipulateAsync(
        this.state.image,
        [{ rotate: 0 }, { flip: ImageManipulator.FlipType.Vertical }],
        { compress: 1, format: ImageManipulator.SaveFormat.PNG, base64: true }
      );

      image = manipResult.base64 || "";
    }

    db.transaction((tx: SQLite.Transaction) => {
      const item: Item = {
        title: this.state.input.title,
        kind: this.state.kind || getKind(this.state.input.title),
        image
      };

      insertItem(tx, item, this.save);
    });
  };

  save = (insertId: number, error: ResultError) => {
    if (error) {
      return;
    }

    if (this.state.input.date) {
      // 日付のデータがある場合ははcalendarに登録する
      db.transaction((tx: SQLite.Transaction) => {
        const item: Calendar = {
          itemId: insertId,
          date: this.state.input.date
        };

        insertCalendar(tx, item, (_, error: ResultError) => {
          if (error) {
            return;
          }

          this.pushCreateSchedule(insertId);
        });
      });
    } else {
      this.pushCreateSchedule(insertId);
    }
  };

  pushCreateSchedule = (itemId: number) => {
    if (this.props.refreshData) {
      this.props.refreshData();
    }

    this.props.navigation.navigate("CreateSchedule", {
      itemId,
      title: this.state.input.title
    });
  };

  onIcons = () => {
    this.props.navigation.navigate("Icons", {
      kind: getKind(this.state.input.title),
      onSelectIcon: (kind: string) => {
        this.props.navigation.navigate("CreatePlan", {
          kind: kind
        });
      },
      onDismiss: () => {
        this.props.navigation.navigate("CreatePlan");
      },
      photo: true
    });
  };

  onCamera = () => {
    this.props.navigation.navigate("Camera", {
      onPicture: (image?: string) => {
        this.props.navigation.navigate("CreatePlan", {
          image
        });
      },
      onDismiss: () => {
        this.props.navigation.navigate("CreatePlan");
      }
    });
  };

  onHome = () => {
    this.props.navigation.goBack(null);
  };

  render() {
    return (
      <Page
        mode="new"
        title={this.state.input.title}
        date={this.state.input.date}
        image={this.state.image}
        kind={this.state.kind}
        suggestList={this.state.suggestList}
        onInput={this.onInput}
        onImage={this.onImage}
        onSave={this.onSave}
        onIcons={this.onIcons}
        onCamera={this.onCamera}
        onHome={this.onHome}
      />
    );
  }
}
