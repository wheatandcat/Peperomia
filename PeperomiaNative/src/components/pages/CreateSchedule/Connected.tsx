import React, { useState, memo, useEffect, useCallback } from 'react';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import { Alert } from 'react-native';
import { Item } from '../../../lib/db/item';
import { ItemDetail } from '../../../lib/db/itemDetail';
import { useDidMount } from '../../../hooks/index';
import { getItemByID } from '../../../lib/item';
import { getItemDetails, updateItemDetail } from '../../../lib/itemDetail';
import Page from '../../templates/CreateSchedule/Page';

type Props = {
  navigation: NavigationScreenProp<NavigationRoute>;
};

type State = {
  item: Item;
  itemDetails: ItemDetail[];
  refresh: string;
};

export default memo((props: Props) => {
  const [state, setState] = useState<State>({
    item: { title: '', kind: '' },
    itemDetails: [],
    refresh: '0',
  });

  const setItemDetails = useCallback(
    (data: ItemDetail[]) => {
      const prioritys = data.map(item => item.priority);
      const uniquePrioritys = prioritys.filter(
        (x: number, i: number, self: number[]) => self.indexOf(x) === i
      );

      // priorityが重複していない
      if (prioritys.length === uniquePrioritys.length) {
        props.navigation.setParams({
          itemDetails: data,
        });
        setState(s => ({
          ...s,
          itemDetails: data,
        }));
      }

      // priorityが重複している場合はid順でpriorityをupdateする
      const itemDetails: ItemDetail[] = data.map(
        (itemDetail: ItemDetail, index: number) => ({
          ...itemDetail,
          priority: index + 1,
        })
      );

      itemDetails.forEach(async (itemDetail, index) => {
        const v = {
          ...itemDetail,
          id: itemDetail.id || '',
          priority: index + 1,
        };

        const ok = await updateItemDetail(null, v);
        if (!ok) {
          Alert.alert('保存に失敗しました');
          return;
        }
      });

      setState(s => ({
        ...s,
        itemDetails: itemDetails,
      }));
    },
    [props.navigation]
  );

  const setItem = useCallback((data: Item) => {
    setState(s => ({
      ...s,
      item: {
        title: data.title,
        kind: data.kind,
      },
    }));
  }, []);

  useDidMount(() => {
    const itemId = props.navigation.getParam('itemId', '1');

    const setItemByItemID = async () => {
      const item = await getItemByID<Item>(null, itemId);
      setItem(item);
    };

    const setItemDetailsByItemID = async () => {
      const itemDetails = await getItemDetails<ItemDetail[]>(null, itemId);
      setItemDetails(itemDetails);
    };

    setItemByItemID();
    setItemDetailsByItemID();
  });

  useEffect(() => {
    const refresh = props.navigation.getParam('refresh', '0');

    const setItemDetailsByItemID = async (itemId: string) => {
      const itemDetails = await getItemDetails<ItemDetail[]>(null, itemId);
      setItemDetails(itemDetails);
    };

    if (refresh !== state.refresh) {
      const itemId = props.navigation.getParam('itemId', '1');
      setItemDetailsByItemID(itemId).then(() => {
        setState(s => ({ ...s, refresh }));
      });
    }
  }, [props.navigation, setItemDetails, state.refresh]);

  const onCreateScheduleDetail = useCallback(() => {
    const itemId = props.navigation.getParam('itemId', '1');

    props.navigation.navigate('CreateScheduleDetail', {
      itemId,
    });
  }, [props.navigation]);

  const onScheduleDetail = useCallback(
    (id: string) => {
      props.navigation.navigate('ScheduleDetail', {
        scheduleDetailId: id,
        priority: state.itemDetails.length + 1,
      });
    },
    [props.navigation, state.itemDetails.length]
  );

  const onFinish = useCallback(() => {
    if (state.itemDetails.length === 0) {
      Alert.alert(
        'まだ予定の設定がありません',
        '本当に完了しますか？',
        [
          {
            text: 'キャンセル',
            style: 'cancel',
          },
          {
            text: '完了する',
            onPress: () => {
              props.navigation.navigate('Home', { refresh: true });
            },
          },
        ],
        { cancelable: false }
      );
    } else {
      props.navigation.navigate('Home', { refresh: true });
    }
  }, [props.navigation, state.itemDetails.length]);

  const onGoBack = useCallback(() => {
    if (state.itemDetails.length === 0) {
      Alert.alert(
        'まだ予定の設定がありません',
        '本当に閉じますか？',
        [
          {
            text: 'キャンセル',
            style: 'cancel',
          },
          {
            text: '閉じる',
            onPress: () => {
              props.navigation.navigate('Home', { refresh: true });
            },
          },
        ],
        { cancelable: false }
      );
    } else {
      props.navigation.navigate('Home', { refresh: true });
    }
  }, [props.navigation, state.itemDetails.length]);

  return (
    <Page
      title={state.item.title}
      kind={state.item.kind}
      data={state.itemDetails}
      onScheduleDetail={onScheduleDetail}
      onCreateScheduleDetail={onCreateScheduleDetail}
      onFinish={onFinish}
      onGoBack={onGoBack}
    />
  );
});
