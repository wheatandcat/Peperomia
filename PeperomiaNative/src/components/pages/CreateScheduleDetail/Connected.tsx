import * as SQLite from "expo-sqlite";
import React, { Component } from "react";
import { NavigationScreenProp, NavigationRoute } from "react-navigation";
import uuidv1 from "uuid/v1";
import { db, ResultError } from "../../../lib/db";
import {
  ItemDetailParam,
  insert as insertItemDetail,
  countByItemId,
  ItemDetail
} from "../../../lib/db/itemDetail";
import { SuggestItem } from "../../../lib/suggest";
import getKind from "../../../lib/getKind";
import {
  Consumer as ItemsConsumer,
  ContextProps
} from "../../../containers/Items";
import Page from "../../templates/CreateScheduleDetail/Page";

export interface State extends ItemDetailParam {
  iconSelected: boolean;
  priority: number;
  suggestList: SuggestItem[];
}

interface Props extends ItemDetailParam {
  navigation: NavigationScreenProp<NavigationRoute>;
}

type PlanProps = Props & Pick<ContextProps, "itemDetails" | "refreshData">;

export default class extends Component<Props> {
  render() {
    return (
      <ItemsConsumer>
        {({ refreshData, itemDetails }: ContextProps) => (
          <Plan
            {...this.props}
            refreshData={refreshData}
            itemDetails={itemDetails}
          />
        )}
      </ItemsConsumer>
    );
  }
}

class Plan extends Component<PlanProps, State> {
  state = {
    title: this.props.title || "",
    kind: this.props.kind || "",
    place: this.props.place || "",
    url: this.props.url || "",
    memo: this.props.memo || "",
    moveMinutes: this.props.moveMinutes || 0,
    iconSelected: false,
    priority: 1,
    suggestList: []
  };

  componentDidMount() {
    const suggestList = (this.props.itemDetails || []).map(itemDetail => ({
      title: itemDetail.title,
      kind: itemDetail.kind
    }));

    this.setState({
      suggestList
    });

    const itemId = this.props.navigation.getParam("itemId", "1");

    db.transaction((tx: SQLite.Transaction) => {
      countByItemId(tx, itemId, this.getCount);
    });
  }

  componentDidUpdate() {
    const kind = this.props.navigation.getParam("kind", "");

    if (!kind) {
      return;
    }

    if (this.state.kind !== kind) {
      this.setState({ kind, iconSelected: true });
    }
  }

  getCount = (count: number, error: ResultError) => {
    if (error) {
      return;
    }

    this.setState({
      priority: count + 1
    });
  };

  onDismiss = () => {
    this.props.navigation.goBack();
  };

  onSave = (
    title: string,
    kind: string,
    place: string,
    url: string,
    memo: string,
    time: number
  ) => {
    const itemId = this.props.navigation.getParam("itemId", "1");

    db.transaction((tx: SQLite.Transaction) => {
      const itemDetail: ItemDetail = {
        itemId,
        title,
        kind,
        place,
        url,
        memo,
        moveMinutes: time,
        priority: this.state.priority
      };

      insertItemDetail(tx, itemDetail, this.save);
    });
  };

  save = (_: number, error: ResultError) => {
    if (error) {
      return;
    }

    const itemId = this.props.navigation.getParam("itemId", "1");

    this.props.navigation.navigate("CreateSchedule", {
      itemId,
      refresh: uuidv1()
    });

    if (this.props.refreshData) {
      this.props.refreshData();
    }
  };

  onIcons = (title: string) => {
    this.props.navigation.navigate("Icons", {
      kind: getKind(title),
      onSelectIcon: (kind: string) => {
        this.props.navigation.navigate("CreateScheduleDetail", {
          kind: kind
        });
      },
      onDismiss: () => {
        this.props.navigation.navigate("CreateScheduleDetail");
      },
      photo: false
    });
  };

  render() {
    return (
      <Page
        title={this.state.title}
        kind={this.state.kind}
        place={this.state.place}
        url={this.state.url}
        memo={this.state.memo}
        time={this.state.moveMinutes}
        suggestList={this.state.suggestList}
        iconSelected={this.state.iconSelected}
        onDismiss={this.onDismiss}
        onSave={this.onSave}
        onIcons={this.onIcons}
      />
    );
  }
}
