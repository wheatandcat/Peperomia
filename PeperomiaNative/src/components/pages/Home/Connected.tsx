import * as SQLite from 'expo-sqlite';
import React, { Component } from 'react';
import { NavigationScreenProp, NavigationRoute } from 'react-navigation';
import { createStackNavigator } from 'react-navigation-stack';
import EStyleSheet from 'react-native-extended-stylesheet';
import { Dimensions, View, Image, AsyncStorage } from 'react-native';
import { Feather } from '@expo/vector-icons';
import uuidv1 from 'uuid/v1';
import theme, { darkMode } from '../../../config/theme';
import { db, ResultError } from '../../../lib/db';
import { Item } from '../../../lib/db/item';
import { delete1st } from '../../../lib/db/item';
import { deleteByItemId as deleteItemDetailByItemId } from '../../../lib/db/itemDetail';
import { deleteByItemId as deleteCalendarByItemId } from '../../../lib/db/calendar';
import {
  Consumer as ItemsConsumer,
  ContextProps,
} from '../../../containers/Items';
import {
  Consumer as ThemeConsumer,
  ContextProps as ThemeContextProps,
} from '../../../containers/Theme';
import Hint from '../../atoms/Hint/Hint';
import Schedule from '../Schedule/Switch';
import EditPlan from '../EditPlan/Connected';
import Page from './Page';

const deviceHeight = Dimensions.get('window').height;
const deviceWidth = Dimensions.get('window').width;

type Props = {
  navigation: NavigationScreenProp<NavigationRoute>;
};

type State = {
  refresh: string;
  mask: boolean;
};

export type PlanProps = Pick<ContextProps, 'items' | 'about' | 'refreshData'> &
  Pick<ThemeContextProps, 'rerendering' | 'onFinishRerendering'> &
  Pick<Props, 'navigation'> & {
    loading: boolean;
    refresh: string;
  };

type PlanState = {
  refresh: string;
};

const LogoTitle = () => (
  <Image
    source={
      darkMode()
        ? require('../../../img/header_dark.png')
        : require('../../../img/header.png')
    }
    style={styles.logo}
    resizeMode="contain"
  />
);

class HomeScreen extends Component<Props, State> {
  static navigationOptions = ({
    navigation,
  }: {
    navigation: NavigationScreenProp<NavigationRoute>;
  }) => {
    const { params = {} } = navigation.state;

    return {
      headerTitle: <LogoTitle />,
      headerStyle: {
        backgroundColor: theme().mode.header.backgroundColor,
      },
      headerRight: (
        <View style={styles.headerRight}>
          <Hint onPress={params.onPushCreatePlan} testID="ScheduleAdd">
            <Feather
              name="plus"
              size={28}
              color={
                darkMode()
                  ? theme().color.highLightGray
                  : theme().color.lightGreen
              }
            />
          </Hint>
        </View>
      ),
    };
  };

  state = {
    refresh: '',
    mask: false,
  };

  async componentDidMount() {
    this.props.navigation.setParams({
      onPushCreatePlan: async () => {
        this.setState({
          mask: false,
        });

        await AsyncStorage.setItem('FIRST_CRAEATE_ITEM', 'true');
        this.props.navigation.navigate('CreatePlan');
      },
    });

    const mask = await AsyncStorage.getItem('FIRST_CRAEATE_ITEM');

    this.setState({
      mask: !mask,
    });
  }

  render() {
    const refresh = this.props.navigation.getParam('refresh', '');

    return (
      <ItemsConsumer>
        {({ items, about, refreshData, itemsLoading }: ContextProps) => (
          <ThemeConsumer>
            {({ rerendering, onFinishRerendering }: ThemeContextProps) => (
              <>
                <HomeScreenPlan
                  loading={Boolean(itemsLoading)}
                  navigation={this.props.navigation}
                  rerendering={rerendering}
                  items={items}
                  about={about}
                  refresh={refresh}
                  refreshData={refreshData}
                  onFinishRerendering={onFinishRerendering}
                />
                {this.state.mask && <View style={styles.mask} />}
              </>
            )}
          </ThemeConsumer>
        )}
      </ItemsConsumer>
    );
  }
}

export class HomeScreenPlan extends Component<PlanProps, PlanState> {
  state = {
    refresh: '',
  };
  componentDidMount() {
    if (this.props.rerendering) {
      this.props.navigation.navigate('ScreenSetting');
      if (this.props.onFinishRerendering) this.props.onFinishRerendering();
    }
  }

  componentDidUpdate() {
    if (this.state.refresh === this.props.refresh) {
      return;
    }

    this.setState({ refresh: this.props.refresh });

    if (this.props.refreshData) {
      this.props.refreshData();
    }
  }

  onDelete = (itemId: string) => {
    db.transaction((tx: SQLite.Transaction) => {
      delete1st(tx, itemId, (_: Item, error: ResultError) => {
        if (error) {
          return;
        }
        deleteCalendarByItemId(tx, Number(itemId), () => null);
        deleteItemDetailByItemId(tx, itemId, this.onRefresh);
      });
    });
  };

  onSchedule = (id: string, title: string) => {
    this.props.navigation.navigate('Schedule', { itemId: id, title });
  };

  onRefresh = () => {
    this.setState({ refresh: uuidv1() });
  };

  render() {
    const items = (this.props.items || []).map((item: Item) => {
      const about = (this.props.about || []).find(
        val => val.itemId === item.id
      );

      return {
        ...item,
        id: String(item.id),
        about: about ? about.about : '',
      };
    });

    return (
      <Page
        data={items}
        loading={this.props.loading}
        onSchedule={this.onSchedule}
        onDelete={this.onDelete}
      />
    );
  }
}

const MainCardNavigator = createStackNavigator(
  {
    Home: {
      screen: HomeScreen,
    },
    Schedule: {
      screen: Schedule,
    },
  },
  {
    defaultNavigationOptions: {
      headerStyle: {
        backgroundColor: theme().mode.header.backgroundColor,
      },
      headerTitleStyle: {
        color: theme().mode.header.text,
      },
      headerTintColor: theme().mode.header.text,
    },
  }
);

export default createStackNavigator(
  {
    MainCardNavigator: {
      screen: MainCardNavigator,
    },
    EditPlan: {
      screen: EditPlan,
    },
  },
  {
    initialRouteName: 'MainCardNavigator',
    headerMode: 'none',
  }
);

const styles = EStyleSheet.create({
  logo: {
    height: 40,
    zIndex: 10,
  },
  headerRight: {
    right: 12,
  },
  mask: {
    position: 'absolute',
    width: deviceWidth,
    height: deviceHeight,
    backgroundColor: 'rgba(0,0,0,0.8)',
  },
});
