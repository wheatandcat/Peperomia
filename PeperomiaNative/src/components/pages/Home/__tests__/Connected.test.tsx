import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Connected from '../Connected';

describe('components/pages/Home/Connected.tsx', () => {
  let wrapper: ShallowWrapper;

  const propsData: any = () => ({});

  beforeEach(() => {
    wrapper = shallow(<Connected {...propsData()} />);
  });

  it('正常に表示されている', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
