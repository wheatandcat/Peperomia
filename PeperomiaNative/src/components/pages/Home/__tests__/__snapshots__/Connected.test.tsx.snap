// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/Home/Connected.tsx 正常に表示されている 1`] = `
<StackNavigator
  initialRouteName="Home"
>
  <Screen
    component={[Function]}
    initialParams={
      Object {
        "refresh": false,
      }
    }
    name="Home"
    options={[Function]}
  />
  <Screen
    component={[Function]}
    name="Schedule"
    options={[Function]}
  />
</StackNavigator>
`;
